pkg/controllers/migration/virtualmachine.go

```diff
@@ -307,44 +312,80 @@ func (h *virtualMachineHandler) preFlightChecks(vm *migration.VirtualMachineImpo
+	logrus.WithFields(logrus.Fields{
+		"name":                         vm.Name,
+		"namespace":                    vm.Namespace,
+		"spec.virtualMachineName":      vm.Spec.VirtualMachineName,
+		"spec.sourceCluster.kind":      vm.Spec.SourceCluster.Kind,
+		"spec.sourceCluster.name":      vm.Spec.SourceCluster.Name,
+		"spec.gracefulShutdownTimeout": vm.Spec.GracefulShutdownTimeout,
+	}).Info("Shutting down guest OS of the source VM")
+	err := vmo.ShutdownGuest(vm)
+	if err != nil {
+		return fmt.Errorf("failed to shutdown the guest OS of the source VM: %w", err)
+	}
+	conds := []common.Condition{
+		{
+			Type:               migration.VirtualMachineShutdownGuest,
+			Status:             v1.ConditionTrue,
+			LastUpdateTime:     metav1.Now().Format(time.RFC3339),
+			LastTransitionTime: metav1.Now().Format(time.RFC3339),
+		},
+	}
+	vm.Status.ImportConditions = util.MergeConditions(vm.Status.ImportConditions, conds)
+	return nil
+}
+
+// triggerPowerOff triggers the power off of the source VM.
+func triggerPowerOff(vm *migration.VirtualMachineImport, vmo VirtualMachineOperations) error {
+	logrus.WithFields(logrus.Fields{
+		"name":                    vm.Name,
+		"namespace":               vm.Namespace,
+		"spec.virtualMachineName": vm.Spec.VirtualMachineName,
+		"spec.sourceCluster.kind": vm.Spec.SourceCluster.Kind,
+		"spec.sourceCluster.name": vm.Spec.SourceCluster.Name,
+	}).Info("Power off the source VM")
+	err := vmo.PowerOff(vm)
+	if err != nil {
+		return fmt.Errorf("failed to power off the source VM: %w", err)
+	}
+	conds := []common.Condition{
+		{
+			Type:               migration.VirtualMachinePoweringOff,
+			Status:             v1.ConditionTrue,
+			LastUpdateTime:     metav1.Now().Format(time.RFC3339),
+			LastTransitionTime: metav1.Now().Format(time.RFC3339),
+		},
+	}
+	vm.Status.ImportConditions = util.MergeConditions(vm.Status.ImportConditions, conds)
+	return nil
+}
+
 func (h *virtualMachineHandler) triggerExport(vm *migration.VirtualMachineImport) error {
 	vmo, err := h.generateVMO(vm)
 	if err != nil {
 		return fmt.Errorf("error generating VMO in trigger export: %v", err)
 	}
 
-	// power off machine
-	if !util.ConditionExists(vm.Status.ImportConditions, migration.VirtualMachinePoweringOff, v1.ConditionTrue) {
-		logrus.WithFields(logrus.Fields{
-			"name":                              vm.Name,
-			"namespace":                         vm.Namespace,
-			"spec.virtualMachineName":           vm.Spec.VirtualMachineName,
-			"spec.sourceCluster.kind":           vm.Spec.SourceCluster.Kind,
-			"spec.sourceCluster.name":           vm.Spec.SourceCluster.Name,
-			"status.importedVirtualMachineName": vm.Status.ImportedVirtualMachineName,
-		}).Info("Power off the source VM")
-		err = vmo.PowerOffVirtualMachine(vm)
-		if err != nil {
-			return fmt.Errorf("failed to power off the source VM: %v", err)
+	// Trigger power off or shutdown guest OS of the source VM.
+	if pointer.BoolDeref(vm.Spec.GracefulShutdown, false) {
```

Comment:
would it be better to just have a timer waiting for `GracefulShutdown` interval and executing `PowerOff` if the timer has run out.

This way it could be one call without having to reconcile later?

Reply:
That would be another approach, but I think it makes it more complicated. Due to the reconcile phase, the whole thing is a simple state machine whose functionality is easy to understand. The waiting time is already set to 5 seconds further down in the code.

```
		logrus.WithFields(logrus.Fields{
			"name":                    vm.Name,
			"namespace":               vm.Namespace,
			"spec.virtualMachineName": vm.Spec.VirtualMachineName,
			"spec.sourceCluster.kind": vm.Spec.SourceCluster.Kind,
			"spec.sourceCluster.name": vm.Spec.SourceCluster.Name,
		}).Info("Waiting for VM to be powered off ...")
		h.importVM.EnqueueAfter(vm.Namespace, vm.Name, 5*time.Second)

		return nil
```

Should we increase this?